### PR TITLE
chore(mesheryctl): add loader to mesheryctl pattern apply

### LINKED ISSUE #5925: [mesheryctl] Add a loader on `mesheryctl pattern apply`
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

#### Current Behavior

When a user applies a pattern using `mesheryctl pattern apply` , no progress or loading is shown to user and after a long time the command exits with success or failure.
#### Desired Behavior

While the command is executing, log `Applying pattern <pattern name>` followed by a loader.
#### Mockups

---

#### Contributor Resources

- ⌨️ [mesheryctl Contributing Guide](https://github.com/meshery/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)
- 🙋🏾🙋🏼 Questions: [Layer5 Discussion Forum](https://discuss.layer5.io) and [Layer5 Community Slack](http://slack.layer5.io)

### PR DESCRIPTION
**Pull request by whois-pk**
Signed-off-by: pk-218 <53428217+pk-218@users.noreply.github.com>

**Description**

This PR fixes #5925 

**Notes for Reviewers**

**[Signed commits](https://github.com/meshery/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [x] Yes, I signed my commits.
 

**Changes Made**

As per the desired behaviour stated in issue #5925, while the command `mesheryctl pattern apply` is executing, the log `Applying pattern <pattern name>` should appear followed by a loader.

- To show pattern name, the top level field `Name` was unmarshaled from its `patternFile` YAML ([apply.go#L227](https://github.com/pk-218/meshery/blob/chore/pk-218/5925/mesheryctl/internal/cli/root/pattern/apply.go#L227))
- To show a loader, a new utilities `NewLoader` and `Stop` were added to the `utils` package ([utils/loader.go](https://github.com/pk-218/meshery/blob/chore/pk-218/5925/mesheryctl/pkg/utils/loader.go#L17))

__Current Behaviour__

Terminal recording 

<a href="https://asciinema.org/a/516263" target="_blank"></a>

Implementation

### CODE REVIEW

**Comment by leecalcote** [comment]
@pk-218 how do you feel about propagating this behavior across other similar commands? Up for this challenge? 

I wonder if you might show what the experience is like for failed provisioning attempts. 

This is a timely subject given the work ensuing on the Messaging Framework and Notification Center for v0.7.0, which squarely aims at improving the Meshery UI side of this same UX. // @Revolyssup @MUzairS15

**Comment by Revolyssup** [comment]
@pk-218  Nice work here.

**Comment by pk-218** [comment]
- @leecalcote Yes! I'd love to add this behaviour to similar commands. Up for the challenge 💪

- For a failure, the loader simply stops on one of the characters of the character set, and the error message is shown as programmed. Using the same `Loader.stop` mutex I can signal a change of character to `✗` like many CLI tools in Linux have.

- This sounds really interesting! I will go through the design document once.

File: mesheryctl/internal/cli/root/pattern/apply.go [hunk 1]
```diff
@@ -222,13 +224,26 @@ mesheryctl pattern apply [pattern-name]
...
+			Services unstructured.Unstructured `yaml:"services"`
+		}
+
+		var p pattern
+		err = yaml.Unmarshal([]byte(patternFile), &p)
+		if err != nil {
+			return errors.Wrap(err, "could not unmarshal pattern file")
+		}
+
+		s := utils.NewLoader("Applying pattern " + p.Name)
```
Comment (nishantwrp) [inline mesheryctl/internal/cli/root/pattern/apply.go:238]: Why not simply use `path.Base(file)` to display name here? We also use it to save the pattern file. 

https://github.com/meshery/meshery/blob/f4fbefd8708d8b535e6bcf6cd80c3d9dbb458187/mesheryctl/internal/cli/root/pattern/apply.go#L114

**Review by nishantwrp — APPROVED** [review]
lgtm, thanks @pk-218! Nice work! Just one nit.

**Review by nishantwrp — CHANGES_REQUESTED** [review]
please fix the failing unit tests too. lmk if you need any help.

**Review by alphaX86 — DISMISSED** [review]
LGTM... But we might want to find a way to fix the output mismatch with test files and with the loader one

**Comment by leecalcote** [comment]
@pk-218 will you itemize each of the other commands that you consider need this same behavior implemented? See https://docs.meshery.io/reference/mesheryctl for reference, if it helps.

**Comment by warunicorn19** [comment]
Hey @pk-218 would you like to talk to about this on the Meshery Dev call today? please feel free to add it to the minutes.

**Comment by pk-218** [comment]
Yes, I was working on that @leecalcote 

Going over all the subcommands -

- [x] `system` 
It contains several verbs `start`, `stop`, `restart`, that take a long time to respond. However, `ViperCompose` used as an instance of `docker-compose` for Viper provides the loader and other dynamic experience to the user.

- [ ]  `perf`
Waiting on a performance test called via `perf apply` can be improved by adding a loader.

- [ ] `mesh`
`mesh deploy` could use a loader, especially when the watch flag (`-w`) is provided.

- [ ] `app`
`app onboard` and `app offboard`

File: mesheryctl/pkg/utils/loader.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	})
+	fmt.Printf("\033[2K")
+}
+
+func (l *Loader) animate() {
+	var m string
+	for i := 0; i < len(symbols); i++ {
+		m = l.Message + " " + symbols[i]
+		Log.Info(m)
+		time.Sleep(time.Millisecond * 200)
```
Comment (leecalcote) [inline mesheryctl/pkg/utils/loader.go:43]: 😬 `time.Sleep()`

Can this loader be event-driven?

Comment (pk-218) [reply mesheryctl/pkg/utils/loader.go:43]: Yup, need to create another channel for that

File: mesheryctl/internal/cli/root/pattern/apply.go [hunk 3]
```diff
@@ -222,13 +223,21 @@ mesheryctl pattern apply [pattern-name]
 			return err
 		}
 
+		pf, err := core.NewPatternFile([]byte(patternFile))
+		if err != nil {
+			return errors.Wrap(err, "could not unmarshal pattern file")
```
Comment (leecalcote) [inline mesheryctl/internal/cli/root/pattern/apply.go:228]:
```suggestion
			return errors.Wrap(err, "Pattern appears invalid. Could not parse successfully")
```

**Review by leecalcote — APPROVED** [review]
Alright. That's tears it. 3 PRs merged within 6 minutes of each other. Who does this @pk-218 guy think he is? He's on 🔥 🔥

**Comment by pk-218** [comment]
Now you've got me gushing 😂
Onto the next one!

**Pull request merged**